File: 303.range-sum-query-immutable/solution.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.N = len(nums)
        self.dp = [0] * self.N

        self.dp[0] = nums[0]
        for i in range(1, self.N):
            self.dp[i] = self.dp[i - 1] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        total = 0
        if (left > 0):
            total -= self.dp[left - 1]
        total += self.dp[right]
        return total


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: 303.range-sum-query-immutable/solution.kt
class NumArray(nums: IntArray) {
    val N = nums.size
    lateinit var dp: IntArray

    init {
        this.dp = IntArray(this.N)
        this.dp[0] = nums[0]
        for (i in 1..this.N - 1) {
            this.dp[i] = this.dp[i - 1] + nums[i]
        }
    }


    fun sumRange(left: Int, right: Int): Int {
        var total = 0
        if (left > 0) {
            total -= this.dp[left - 1]
        }
        return total + this.dp[right]
    }

}

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */
